jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/IOResources.java

```diff
@@ -314,19 +352,23 @@ public static void copy(Resource resource, Content.Sink sink, ByteBufferPool buf
         // Directly write the byte array if the resource is a MemoryResource.
         if (resource instanceof MemoryResource memoryResource)
         {
-            byte[] bytes = memoryResource.getBytes();
-            ByteBuffer byteBuffer = ByteBuffer.wrap(bytes);
-            if (first >= 0)
-                byteBuffer.position((int)first);
-            if (length >= 0)
-                byteBuffer.limit((int)(byteBuffer.position() + length));
-            sink.write(true, byteBuffer, callback);
+            sink.write(true, BufferUtil.slice(ByteBuffer.wrap(memoryResource.getBytes()), Math.toIntExact(first), Math.toIntExact(length)), callback);
             return;
         }
 
         // Fallback to Content.Source.
-        Content.Source source = asContentSource(resource, bufferPool, bufferSize, direct, first, length);
-        Content.copy(source, sink, callback);
+        try
+        {
+            InputStream inputStream = resource.newInputStream();
+            if (inputStream == null)
+                throw new IllegalArgumentException("Resource does not support InputStream: " + resource);
+            Content.Source source = Content.Source.from(new ByteBufferPool.Sized(bufferPool, direct, bufferSize), inputStream, first, length);
+            Content.copy(source, sink, callback);
+        }
+        catch (IOException e)
+        {
+            throw new RuntimeIOException(e);
```

Comment:
Can't throw, but fail the callback instead.

Reply:
Changed.